manta-crypto/src/dalek/ed25519.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Dalek Cryptography `ed25519` Backend
+
+use crate::{
+    rand::{CryptoRng, Rand, RngCore},
+    signature::{
+        MessageType, RandomnessType, Sign, SignatureType, SigningKeyType, Verify, VerifyingKeyType,
+    },
+};
+use core::marker::PhantomData;
+use manta_util::AsBytes;
+
+pub use ed25519_dalek::*;
+
+/// Converts `bytes` into a [`SecretKey`].
+#[inline]
+pub fn secret_key_from_bytes(bytes: [u8; SECRET_KEY_LENGTH]) -> SecretKey {
+    match SecretKey::from_bytes(&bytes) {
+        Ok(secret_key) => secret_key,
+        _ => {
+            unreachable!("We are guaranteed the correct number of bytes from `SECRET_KEY_LENGTH`.")
+        }
+    }
+}
+
+/// Clones the `secret_key` by serializing and then deserializing.
+#[inline]
+pub fn clone_secret_key(secret_key: &SecretKey) -> SecretKey {
+    secret_key_from_bytes(secret_key.to_bytes())
+}
+
+/// Generates a [`Keypair`] from `secret_key`.
+#[inline]
+pub fn keypair(secret_key: &SecretKey) -> Keypair {
+    Keypair {
+        public: secret_key.into(),
+        secret: clone_secret_key(secret_key),
+    }
+}
+
+/// Generates a [`SecretKey`] from `rng`.
+#[inline]
+pub fn generate_secret_key<R>(rng: &mut R) -> SecretKey
+where
+    R: CryptoRng + RngCore,
+{
+    secret_key_from_bytes(rng.gen())
+}
+
+/// Generates a [`Keypair`] from `rng`.
+#[inline]
+pub fn generate_keypair<R>(rng: &mut R) -> Keypair
+where
+    R: CryptoRng + RngCore,
+{
+    let secret_key = generate_secret_key(rng);
+    Keypair {
+        public: (&secret_key).into(),
+        secret: secret_key,
+    }
+}
+
+/// Edwards Curve Signature Scheme for the `Curve25519` Elliptic Curve
+#[derive(derivative::Derivative)]
+#[derivative(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
```

Comment:
Why should we derive `Hash`, `Ord`, `PartialEq`, and `PartialOrd` for a signature function `ed25519`?

Reply:
It's a marker struct so it should derive all of the standard traits. We do this to make sure that these structures are compatible with interfaces that do not use correct derive semantics (like the standard Rust ones) and pose _extra_ unnecessary constraints on their types. Think something like the following:

```rust
#[derive(Clone, PartialEq)]
struct Object<T, S> {
    inner: T,
    __: PhantomData<S>,
}
```

This would also require that `S: Clone` or `S: PartialEq` whenever `Object<T, S>` needs to, which is the incorrect semantics. By setting marker structs to have all the standard traits we avoid this footgun for people who don't use the correct `derive` semantics because in this case if `S` were the `Ed25519` struct above, it would actually have implementations for all of those traits.